openwisp_monitoring/monitoring/base/models.py

```diff
@@ -213,13 +213,18 @@ def write(
             database=database,
             retention_policy=retention_policy,
         )
-        timeseries_write.delay(name=self.key, values=values, **options)
-        post_metric_write.send(**signal_kwargs)
         # check can be disabled,
         # mostly for automated testing and debugging purposes
-        if not check:
-            return
-        self.check_threshold(value, time, retention_policy, send_alert)
+        if check:
+            options['threshold_dict'] = {
+                'value': value,
+                'time': time,
+                'retention_policy': retention_policy,
+                'send_alert': send_alert,
+            }
+            options['metric_pk'] = self.pk
+        timeseries_write.delay(name=self.key, values=values, **options)
+        post_metric_write.send(**signal_kwargs)
```

Comment:
I think now is a good time to move this signal to the background task and execute it after the metric has been written.
We should also update the related docs to explain it will fire only after a successful metric write.
The way it's implemented now is really not very useful.

Reply:
Agreed :+1: